compiler/qsc/benches/rca.rs

```diff
@@ -0,0 +1,163 @@
+            });
+        },
+    );
+}
+
+pub fn teleport(c: &mut Criterion) {
+    c.bench_function(
+        "Perform Runtime Capabilities Analysis (RCA) on teleport sample",
+        |b| {
+            // First, compile and analyze the packages included by default (core & std).
+            let mut compilation_context = CompilationContext::new();
+            compilation_context.analyze_all();
+
+            // Now, update the compilation with the sample, and analyze only the updated package.
+            compilation_context.update_compilation(TELEPORT);
+            b.iter(|| {
+                compilation_context.analyze_open_package();
+            });
+        },
+    );
+}
+
+pub fn deutsch_jozsa(c: &mut Criterion) {
+    c.bench_function(
+        "Perform Runtime Capabilities Analysis (RCA) on Deutsch-Jozsa sample",
+        |b| {
+            // First, compile and analyze the packages included by default (core & std).
+            let mut compilation_context = CompilationContext::new();
+            compilation_context.analyze_all();
+
+            // Now, update the compilation with the sample, and analyze only the updated package.
+            compilation_context.update_compilation(DEUTSCHJOZSA);
+            b.iter(|| {
+                compilation_context.analyze_open_package();
+            });
+        },
+    );
+}
+
+pub fn large_file(c: &mut Criterion) {
+    c.bench_function(
+        "Perform Runtime Capabilities Analysis (RCA) on large file sample",
+        |b| {
+            // First, compile and analyze the packages included by default (core & std).
+            let mut compilation_context = CompilationContext::new();
+            compilation_context.analyze_all();
+
+            // Now, update the compilation with the sample, and analyze only the updated package.
+            compilation_context.update_compilation(LARGE);
+            b.iter(|| {
+                compilation_context.analyze_open_package();
+            });
+        },
+    );
+}
+
+struct CompilationContext {
+    compiler: Compiler,
+    lowerer: Lowerer,
+    fir_store: PackageStore,
+    compute_properties: Option<PackageStoreComputeProperties>,
+}
+
+impl CompilationContext {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    fn analyze_all(&mut self) {
+        let analyzer = Analyzer::init(&self.fir_store);
+        let compute_properties = analyzer.analyze_all();
+        self.compute_properties = Some(compute_properties);
+    }
+
+    fn analyze_open_package(&mut self) {
+        let Some(compute_properties) = &mut self.compute_properties else {
+            panic!("cannot analyze open package if the other packages have not been analyzed");
+        };
+
+        // Clear the compute properties of the open package.
+        let open_package_id = map_hir_package_to_fir(self.compiler.package_id());
+        let package_compute_properties = compute_properties.get_mut(open_package_id);
+        package_compute_properties.clear();
```

Comment:
random question: if we haven't yet computed the properties for the open package, why do they need to be cleared?

Reply:
Oh, never mind, it's because of the repeated benchmarking.